File: internal/execution/execution_test.go
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2026 Steadybit GmbH

package execution_test

import (
	"context"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"

	"github.com/steadybit/cli/internal/execution"
	"github.com/steadybit/cli/internal/output"
	"github.com/steadybit/cli/internal/platformtest"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

var ctx = context.Background()

const run = `{"id":42,"key":"ADM-1","state":"COMPLETED","steps":[
 {"stepType":"wait"},
 {"stepType":"action","actionId":"com.steadybit.extension_jmeter.run","targetExecutions":[
   {"id":"te-1","name":"host-a","artifacts":["report.zip","log.txt"]},
   {"id":"te-2","name":"host-b","artifacts":["report.zip"]}]},
 {"stepType":"service-validation","customLabel":"shop is healthy","validations":[
   {"stepType":"action","targetExecutions":[{"id":"te-3","name":"check","artifacts":["result.json"]}]}]}]}`

func TestCollectsArtifactsOfActionsAndServiceValidations(t *testing.T) {
	doc, err := output.ParseDocument([]byte(run))
	require.NoError(t, err)

	assert.Equal(t, []execution.Artifact{
		{Step: "com.steadybit.extension_jmeter.run", Target: "host-a", TargetExecutionID: "te-1", ArtifactID: "report.zip"},
		{Step: "com.steadybit.extension_jmeter.run", Target: "host-a", TargetExecutionID: "te-1", ArtifactID: "log.txt"},
		{Step: "com.steadybit.extension_jmeter.run", Target: "host-b", TargetExecutionID: "te-2", ArtifactID: "report.zip"},
		{Step: "shop is healthy", Target: "check", TargetExecutionID: "te-3", ArtifactID: "result.json"},
	}, execution.Collect(doc.Value()))
}

// The platform leaves the steps out unless asked, and with them every artifact.
func TestAsksForTheSteps(t *testing.T) {
	p := platformtest.New(t)
	p.Reply("GET /api/experiments/executions/42", platformtest.Reply{JSON: map[string]any{"id": 42, "state": "RUNNING"}})

	out, err := platformtest.Stdout(t, func() error { return execution.Get(ctx, p.Client, execution.GetOptions{ID: 42}) })

	require.NoError(t, err)
	assert.Equal(t, "id: 42\nstate: RUNNING\n\n", out)
	assert.Equal(t, []string{"steps"}, p.Requests("GET /api/experiments/executions/42")[0].Query["fields"])
}

func TestCancel(t *testing.T) {
	p := platformtest.New(t)
	p.Reply("POST /api/experiments/executions/42/cancel", platformtest.Reply{Status: http.StatusAccepted})
	p.Reply("POST /api/experiments/executions/43/cancel", platformtest.Reply{Status: http.StatusOK})
	p.Reply("POST /api/experiments/executions/44/cancel", platformtest.Reply{Status: http.StatusNotFound})

	out, err := platformtest.Stdout(t, func() error {
		if err := execution.Cancel(ctx, p.Client, 42); err != nil {
			return err
		}
		return execution.Cancel(ctx, p.Client, 43)
	})
	require.NoError(t, err)
	assert.Equal(t, "Experiment run 42 is being canceled.\nExperiment run 43 has already ended.\n", out)
	assert.EqualError(t, execution.Cancel(ctx, p.Client, 44), "Experiment run 44 not found.")
}

func TestPropertyValues(t *testing.T) {
	p := platformtest.New(t)
	p.Reply("POST /api/experiments/executions/42/properties/*/set", platformtest.Reply{})
	p.Reply("POST /api/experiments/executions/42/properties/*/add", platformtest.Reply{})
	set := func(values []string, asJSON bool) string {
		_, err := platformtest.Stdout(t, func() error {
			return execution.SetProperty(ctx, p.Client, execution.PropertyOptions{ID: 42, Key: "k", Values: values, JSON: asJSON})
		})
		require.NoError(t, err)
		requests := p.Requests("POST /api/experiments/executions/42/properties/k/set")
		return string(requests[len(requests)-1].Body)
	}

	assert.Equal(t, `"0042"`, set([]string{"0042"}, false), "a number-like value stays a string")
	assert.Equal(t, `["a","b"]`, set([]string{"a", "b"}, false))
	assert.Equal(t, `7`, set([]string{"7"}, true))
	assert.Equal(t, `0`, set([]string{"0"}, true), "falsy values are sent")
	assert.Equal(t, `false`, set([]string{"false"}, true))
	assert.Equal(t, `""`, set([]string{""}, false))

	err := execution.SetProperty(ctx, p.Client, execution.PropertyOptions{ID: 42, Key: "k", Values: []string{"seven"}, JSON: true})
	assert.ErrorContains(t, err, "'seven' is not valid JSON")
	assert.EqualError(t, execution.AddProperty(ctx, p.Client, execution.PropertyOptions{ID: 42, Key: "k", Values: []string{"a", "b"}}),
		"Adding to a list property takes exactly one --value.")
}

func TestPropertyErrorsNameTheProperty(t *testing.T) {
	p := platformtest.New(t)
	p.Reply("POST /api/experiments/executions/42/properties/locked/set", platformtest.Reply{Status: 422, JSON: map[string]any{"title": "not editable"}})

	err := execution.SetProperty(ctx, p.Client, execution.PropertyOptions{ID: 42, Key: "locked", Values: []string{"x"}})

	assert.ErrorContains(t, err, "Failed to set property locked of experiment run 42: ")
	assert.ErrorContains(t, err, "not editable")
}

func TestDownloadsEveryArtifactIntoADirectoryPerTarget(t *testing.T) {
	p := platformtest.New(t)
	p.Reply("GET /api/experiments/executions/42", platformtest.Reply{Body: run})
	p.Handle("GET /api/experiments/executions/42/artifacts/*/*", func(r platformtest.Request) platformtest.Reply {
		parts := strings.Split(r.Path, "/")
		return platformtest.Reply{Body: "content of " + parts[6] + "/" + parts[7]}
	})
	dir := t.TempDir()

	_, err := platformtest.Stdout(t, func() error {
		return execution.Download(ctx, p.Client, execution.DownloadOptions{ID: 42, Directory: dir})
	})

	require.NoError(t, err)
	for _, f := range []string{"te-1/report.zip", "te-1/log.txt", "te-2/report.zip", "te-3/result.json"} {
		content, err := os.ReadFile(filepath.Join(dir, f))
		require.NoError(t, err)
		assert.Equal(t, "content of "+f, string(content))
	}
}

func TestDownloadToASingleFile(t *testing.T) {
	p := platformtest.New(t)
	p.Reply("GET /api/experiments/executions/42", platformtest.Reply{Body: run})
	p.Reply("GET /api/experiments/executions/42/artifacts/te-1/log.txt", platformtest.Reply{Body: "the log"})
	file := filepath.Join(t.TempDir(), "my.log")

	_, err := platformtest.Stdout(t, func() error {
		return execution.Download(ctx, p.Client, execution.DownloadOptions{ID: 42, Artifact: "log.txt", Output: file})
	})

	require.NoError(t, err)
	content, _ := os.ReadFile(file)
	assert.Equal(t, "the log", string(content))
	err = execution.Download(ctx, p.Client, execution.DownloadOptions{ID: 42, Artifact: "report.zip", Output: file})
	assert.EqualError(t, err, "2 artifacts match, but --output takes exactly one. Narrow it down with --artifact and --target-execution.")
	err = execution.Download(ctx, p.Client, execution.DownloadOptions{ID: 42, Artifact: "nope"})
	assert.EqualError(t, err, "No matching artifacts found in experiment run 42.")
}

func TestNeverLetsAnIdStepOutOfADirectory(t *testing.T) {
	for _, id := range []string{"..", ".", "", "../..", "a/../..", "/"} {
		assert.Equal(t, "_", execution.PathSegment(id), id)
	}
	assert.Equal(t, "evil", execution.PathSegment("../../evil"))
	assert.Equal(t, "report.zip", execution.PathSegment("report.zip"))
}

func TestWatchPrintsChangesUntilTheRunEnds(t *testing.T) {
	p := platformtest.New(t)
	p.Reply("GET /api/experiments/ADM-1/executions", platformtest.Reply{JSON: map[string]any{"executions": []any{map[string]any{"id": 41}, map[string]any{"id": 42}}}})
	var polls int
	p.Handle("GET /api/experiments/executions/42", func(platformtest.Request) platformtest.Reply {
		polls++
		state, step := "RUNNING", "RUNNING"
		if polls > 1 {
			state, step = "FAILED", "FAILED"
		}
		return platformtest.Reply{JSON: map[string]any{"id": 42, "key": "ADM-1", "state": state, "reason": "check failed", "steps": []any{
			map[string]any{"stepType": "action", "actionId": "http-check", "state": step, "targetExecutions": []any{map[string]any{"state": step}}},
		}}}
	})

	out, err := platformtest.Stdout(t, func() error {
		return execution.Watch(ctx, p.Client, execution.WatchOptions{Key: "ADM-1", Interval: time.Millisecond})
	})

	assert.EqualError(t, err, "Experiment ADM-1 (#42) failed, reason: check failed")
	assert.Equal(t, "Experiment ADM-1 run #42: running\n  step 1/1 http-check: running (targets 0/1)\nExperiment ADM-1 run #42: failed\n  step 1/1 http-check: failed (targets 1/1)\n", out)
	assert.Empty(t, p.Requests("POST /api/experiments/executions/42/cancel"), "watching never cancels")
}

File: src/execution/execution.test.ts
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2026 Steadybit GmbH

import fs from 'node:fs/promises';
import path from 'node:path';
import { describe, expect, it, vi } from 'vitest';
import { HttpResponse } from 'msw';
import { respondTo } from '../mocks/recorder.ts';
import { getTempDir } from '../mocks/tempFiles.ts';
import { cancel } from './cancel.ts';
import { addProperty, setProperty } from './property.ts';
import { collectArtifacts, downloadArtifacts, listArtifacts, pathSegment } from './artifacts.ts';
import { getExecution } from './get.ts';
import type { Execution } from './api.ts';

const EXECUTION = {
  id: 42,
  key: 'ADM-1',
  state: 'COMPLETED',
  steps: [
    { stepType: 'wait', id: 'w' },
    {
      stepType: 'action',
      actionId: 'com.steadybit.extension_jmeter.run',
      targetExecutions: [
        { id: 'te-1', name: 'host-a', artifacts: ['report.zip', 'log.txt'] },
        { id: 'te-2', name: 'host-b', artifacts: ['report.zip'] },
      ],
    },
    {
      stepType: 'service-validation',
      customLabel: 'shop is healthy',
      validations: [
        { stepType: 'action', targetExecutions: [{ id: 'te-3', name: 'check', artifacts: ['result.json'] }] },
      ],
    },
  ],
} as unknown as Execution;

describe('execution', () => {
  describe('get', () => {
    it('prints the run as YAML by default', async () => {
      respondTo('get', '/api/experiments/executions/42', () => ({ json: { id: 42, state: 'RUNNING' } }));
      const logSpy = vi.spyOn(console, 'log');

      await getExecution({ id: 42 });

      expect(logSpy).toHaveBeenCalledWith('id: 42\nstate: RUNNING\n');
    });

    it('reports a run that does not exist', async () => {
      respondTo('get', '/api/experiments/executions/43', () => ({ status: 404 }));

      await expect(getExecution({ id: 43 })).rejects.toThrow('Experiment run 43 not found.');
    });
  });

  describe('cancel', () => {
    it('reports a cancel that was accepted', async () => {
      const requests = respondTo('post', '/api/experiments/executions/42/cancel', () => ({ status: 202 }));
      const logSpy = vi.spyOn(console, 'log');

      await cancel({ id: 42 });

      expect(requests).toHaveLength(1);
      expect(logSpy).toHaveBeenCalledWith('Experiment run %s is being canceled.', 42);
    });

    it('reports a run that had already ended', async () => {
      respondTo('post', '/api/experiments/executions/42/cancel', () => ({ status: 200 }));
      const logSpy = vi.spyOn(console, 'log');

      await cancel({ id: 42 });

      expect(logSpy).toHaveBeenCalledWith('Experiment run %s has already ended.', 42);
    });

    it('reports a run that does not exist', async () => {
      respondTo('post', '/api/experiments/executions/43/cancel', () => ({ status: 404 }));

      await expect(cancel({ id: 43 })).rejects.toThrow('Experiment run 43 not found.');
    });
  });

  describe('property', () => {
    it('sends a single value as a string, even when it looks like a number', async () => {
      const requests = respondTo('post', '/api/experiments/executions/42/properties/ticket/set', () => ({}));

      await setProperty({ id: 42, key: 'ticket', value: ['0042'] });

      expect(requests[0].body).toBe('0042');
    });

    it('sends several values as a list', async () => {
      const requests = respondTo('post', '/api/experiments/executions/42/properties/tickets/set', () => ({}));

      await setProperty({ id: 42, key: 'tickets', value: ['SHOP-1', 'SHOP-2'] });

      expect(requests[0].body).toEqual(['SHOP-1', 'SHOP-2']);
    });

    it('parses values as JSON on request', async () => {
      const requests = respondTo('post', '/api/experiments/executions/42/properties/score/set', () => ({}));

      await setProperty({ id: 42, key: 'score', value: ['7'], json: true });

      expect(requests[0].body).toBe(7);
    });

    it('rejects a value that is not JSON with --json', async () => {
      await expect(setProperty({ id: 42, key: 'score', value: ['seven'], json: true })).rejects.toThrow(
        "'seven' is not valid JSON"
      );
    });

    it('adds a value to a list property', async () => {
      const requests = respondTo('post', '/api/experiments/executions/42/properties/tickets/add', () => ({}));
      const logSpy = vi.spyOn(console, 'log');

      await addProperty({ id: 42, key: 'tickets', value: ['SHOP-3'] });

      expect(requests[0].body).toBe('SHOP-3');
      expect(logSpy).toHaveBeenCalledWith('Property %s of experiment run %s updated.', 'tickets', 42);
    });

    it('surfaces the platform validation error', async () => {
      respondTo('post', '/api/experiments/executions/42/properties/locked/set', () => ({
        status: 422,
        json: { title: 'Property locked is not editable in an execution' },
      }));

      await expect(setProperty({ id: 42, key: 'locked', value: ['x'] })).rejects.toThrow(
        'Property locked is not editable in an execution'
      );
    });
  });

  describe('artifacts', () => {
    it('collects artifacts of action steps and service validations', () => {
      expect(collectArtifacts(EXECUTION)).toEqual([
        {
          step: 'com.steadybit.extension_jmeter.run',
          target: 'host-a',
          targetExecutionId: 'te-1',
          artifactId: 'report.zip',
        },
        {
          step: 'com.steadybit.extension_jmeter.run',
          target: 'host-a',
          targetExecutionId: 'te-1',
          artifactId: 'log.txt',
        },
        {
          step: 'com.steadybit.extension_jmeter.run',
          target: 'host-b',
          targetExecutionId: 'te-2',
          artifactId: 'report.zip',
        },
        { step: 'shop is healthy', target: 'check', targetExecutionId: 'te-3', artifactId: 'result.json' },
      ]);
    });

    it('says so when a run has no artifacts', async () => {
      respondTo('get', '/api/experiments/executions/42', () => ({ json: { id: 42, steps: [] } }));
      const logSpy = vi.spyOn(console, 'log');

      await listArtifacts({ id: 42 });

      expect(logSpy).toHaveBeenCalledWith('Experiment run %s has no artifacts.', 42);
    });

    it('downloads every artifact into a directory per target execution', async () => {
      respondTo('get', '/api/experiments/executions/42', () => ({ json: EXECUTION as any }));
      respondTo(
        'get',
        '/api/experiments/executions/42/artifacts/:target/:artifact',
        ({ url }) => new HttpResponse(`content of ${url.pathname.split('/').slice(-2).join('/')}`)
      );
      const directory = path.join(getTempDir(), 'artifacts');

      await downloadArtifacts({ id: 42, directory });

      await expect(fs.readFile(path.join(directory, 'te-1', 'report.zip'), 'utf8')).resolves.toBe(
        'content of te-1/report.zip'
      );
      await expect(fs.readFile(path.join(directory, 'te-2', 'report.zip'), 'utf8')).resolves.toBe(
        'content of te-2/report.zip'
      );
      await expect(fs.readFile(path.join(directory, 'te-3', 'result.json'), 'utf8')).resolves.toBe(
        'content of te-3/result.json'
      );
    });

    it('writes a single artifact to --output', async () => {
      respondTo('get', '/api/experiments/executions/42', () => ({ json: EXECUTION as any }));
      respondTo('get', '/api/experiments/executions/42/artifacts/te-1/log.txt', () => new HttpResponse('the log'));
      const output = path.join(getTempDir(), 'my.log');

      await downloadArtifacts({ id: 42, directory: '.', artifact: 'log.txt', output });

      await expect(fs.readFile(output, 'utf8')).resolves.toBe('the log');
    });

    it('refuses --output when several artifacts match', async () => {
      respondTo('get', '/api/experiments/executions/42', () => ({ json: EXECUTION as any }));

      await expect(
        downloadArtifacts({ id: 42, directory: '.', artifact: 'report.zip', output: 'report.zip' })
      ).rejects.toThrow('2 artifacts match, but --output takes exactly one.');
    });

    it('keeps downloads inside the directory whatever the artifact is called', async () => {
      respondTo('get', '/api/experiments/executions/42', () => ({
        json: {
          id: 42,
          steps: [{ stepType: 'action', targetExecutions: [{ id: 'te-9', artifacts: ['../../evil'] }] }],
        },
      }));
      respondTo('get', '/api/experiments/executions/42/artifacts/:target/:artifact', () => new HttpResponse('x'));
      const directory = path.join(getTempDir(), 'contained');

      await downloadArtifacts({ id: 42, directory });

      await expect(fs.readFile(path.join(directory, 'te-9', 'evil'), 'utf8')).resolves.toBe('x');
    });

    it('never lets an id step out of a directory', () => {
      expect(['..', '.', '', '../..', 'a/../..'].map(pathSegment)).toEqual(['_', '_', '_', '_', '_']);
      expect(pathSegment('report.zip')).toBe('report.zip');
    });

    it('reports when nothing matches', async () => {
      respondTo('get', '/api/experiments/executions/42', () => ({ json: EXECUTION as any }));

      await expect(downloadArtifacts({ id: 42, directory: '.', artifact: 'nope' })).rejects.toThrow(
        'No matching artifacts found in experiment run 42.'
      );
    });
  });
});
